### DISCUSSION TITLE
How do I install dependencies?

### DISCUSSION THREAD

**Discussion by rdmchr** [vercel/turborepo#872] [Help] [answered]
I just started using Turborepo and I was wondering how I can install dependencies? Let's say I want to use TailwindCSS in my `web` app, how would I install that dependency just in that app?

**Comment by rdmchr** [comment, marked as answer]
`npm install <package name> -w=<workspace to install in>` appears to be the correct way. Thanks, Rafaeltab for helping me on Discord.

**Reply by DanielBailey-web** [reply]
Just for anyone who finds this looking for the yarn solution, use
`yarn workspace <workspace name> add <package name>`

**Reply by wiredmatt** [reply]
I have an entry for each package in `scripts` in my root package.json:

```json
{
"<workspace name>": "yarn workspace <workspace name>"
"web": "yarn workspace web"
}
```

So then I can just do

`yarn <workspace name> add <package>`
`yarn web add motion`

**Reply by prometheas** [reply]
As someone just getting started with Turborepo, I can genuinely report this was way too difficult to figure out how to accomplish, given how common a task this is. Really ought to be in the docs.

**Reply by DesignByOnyx** [reply]
I usually `cd` into the folder I want and just do a normal `npm install` or `yarn add`. People might complain about the extra command, but at least you get auto-complete with `cd` and run less of a risk of mis-typing something - and there's no new CLI flags to learn. This is particularly useful for onboarding people who are new to monorepos - you just tell them to treat each package like they're used to doing.

**Comment by ArnabXD** [comment]
I am facing same issue with yarn

That doesn't work, for me atleast.

```sh
yarn workspace w1 add some-package 
```

```
yarn workspace v1.22.15
yarn add v1.22.15
info No lockfile found.
error Running this command will add the dependency to the workspace root rather than the workspace itself, which might not be what you want - if you really meant it, make it explicit by running this command again with the -W flag (or --ignore-workspace-root-check).
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
error Command failed.
Exit code: 1
```

I have lockfile in root, (`--no-lockfile` doesn't help either)

**Reply by DanielBailey-web** [reply]
Try and upgrade yarn to 1@latest maybe?

Also, make sure that you have a:
```
  "name": "w1",
```

in your package.json

**Reply by ArnabXD** [reply]
yarn is already on v1.22.15 and my workspace names are right because I can run workspace specific scripts. e.g: 

```bash
yarn workspace w1 run dev
```

**Comment by kevinswarner** [comment]
I am a little confused on best practices for common dev dependencies. For example, let's say you are using Tailwind CSS on multiple workspace applications in your monorepo. Should the dependencies be added to the root monorepo level or to each individual workspace level. I see that the default Turborepo setup has prettier installed at the root level, along with turbo. I am just looking for some guidance and context about this topic. Thanks!

**Reply by lodmfjord** [reply]
I normally would install packages like that as devDependency and peerDependency and then add the package as a dependency in the web app that is importing all the components using tailwind.

**Reply by ackvf** [reply]
@lodmfjord Wait what?

**Comment by kierrereeg** [comment]
I am having an issue with the pnpm version of this 

i see in the docs they say do

pnpm add 'package' --filter 'workspace'

I have a simple setup at the moment and its just giving me this back in the command line:
No projects matched the filters in "[my directory path]" 

I have triple checked the pnpm-workspace.yaml file and compared it to the files in my root, i have gone through this process multiple times and it always does the same thing

I wonder if anyone else is experiencing similar issues?

FYI I am running a GitHub codespace instance  and when i uname -v i get 
#77~18.04.1-Ubuntu SMP Wed Mar 30 15:36:02 UTC 2022

**Reply by kingcw** [reply]
pnpm add package-name --workspace

**Reply by kierrereeg** [reply]
Thanks, tried it, doesn't work, and also doesn't give the no projects filters error, so idk my gut is telling me that one is closer... I have been over and over and over these docs and its not really clear, well its clear what they say, its just not working...

**Reply by kierrereeg** [reply]
FYI to anyone who is seeing this, i also reinstalled everything and did it from the ground up, on pnpm package at a time, everything works fine, including the turbo builds, until i try to tell it the 

pnpm add        whatever      --filter       file name

as described in the Turborepo docs, 

I can install with a -w but its just janky and its not working correctly, idk been trying different things for days at this point....

**Reply by kierrereeg** [reply]
I have also looked at a version of turborepo that was built via the npx create-turbo@latest and my files are the same....

I have no idea why the pnpm workspaces are not working correctly at this point

**Reply by kingcw** [reply]
I don't understand what you are doing.  -w is to add package to project root, but you are using workspaces, are you using project root for a real project?   put that project in an app folder, and add app/* to pnpm-workspace.yaml, cd into that project folder, and use --workspace command

**Reply by kierrereeg** [reply]
I was able to figure it out the correct command is 

pnpm add (package name here) --filter (workspace name here) 

as they say on the Turborepo website, and is the command I had been using the whole time, the fix was to go into the workspace file folders I had created and run the |      pnpm init      | command, this gave the workspaces I declared in the |    pnpm-workspace.yaml     
     | file 
a package.json file to read and thus, i assume, recognize that the workspace existed in the first place, so for pnpm you have to declare the workspace structure in the|.    pnpm-workspaces.yaml      | file and then run |       pnpm init         | on each of the workspace file folders you created to match the structure of the   |    pnpm-workspaces.yaml  |  file, now pnpm can recognize the workspace because of the package.json in each file folder I assume, anyways its fixed now and everyone can hopefully stumble into this thread and solve this, thanks for your help!

**Reply by kierrereeg** [reply]
And I can use that above mentioned command from anywhere in the monorepo with regard to where I am in the file structure of the terminal, I haven't tested every folder but it works in a couple different ones....

**Comment by anricoj1** [comment]
Are there any good resources for this? I can't seem to find any information on dependency management in the Docs. The accepted answer is somewhat helpful, but I don't really understand how it works

**Reply by anricoj1** [reply]
For anyone coming into this thread looking for more answers, a new issue was opened on this to have more info added to the docs.
https://github.com/vercel/turborepo/issues/2060

**Comment by paulwongx** [comment]
For anyone else reading this, I think [this blog](https://blog.andrewbrey.com/2022-10-12-your-monorepo-dependencies-are-asking-for-trouble/) is really helpful as I encounted an issue with different dependency versions which broke my whole app. Namely, upgrading `@types/react` in one workspace broke all the other workspaces. If I did it right from the start, it would've been less of a headache. 

Essentially.... for dependencies that are global across your app, install it globally so it shows up in your root `package.json`. For example 
```cli
$ yarn add next react react-dom -W
$ yarn add -D @tpyes/react @types/node @types/react-dom -W
```
\* The `-W` flag is needed to force installing in the root since by default it warns you if you don't have the `-W` flag
** `-W` is short for `--ignore-workspace-root-check` 

You do not need to add it in the local `package.json` again. The compiler will look up your monorepo tree and find the root `package.json` with the dependency and it's version.

For packages that are local to a package, install it with
```cli
$ yarn workspace <workspace_name> add <package_name>

# For example
$ yarn workspace @core/ui add framer-motion
```
